homeassistant/components/bluetooth/match.py

```diff
@@ -2,12 +2,12 @@
 from __future__ import annotations
 
 from dataclasses import dataclass
-import fnmatch
 from typing import TYPE_CHECKING, Final, TypedDict
 
 from lru import LRU  # pylint: disable=no-name-in-module
 
 from homeassistant.loader import BluetoothMatcher, BluetoothMatcherOptional
+from homeassistant.util.fnmatch import memorized_fnmatch
```

Comment:
Instead of having a memorized fnmatch that memorizes across any consumer, let’s make a factory method so ie, Bluetooth can create it’s own memorized matcher. Otherwise we could see degraded functionality if some random custom component starts to fill up the LRU cache with stuff.

Reply:
I'd be a bit worried about making an object since the `lru_cache` will hold a strong reference to `self`. If someone creates a lot of them it could start to be an issue.

Maybe I should make a copy in bluetooth instead and let the other ones share this since they aren't anywhere near the impact

Reply:
I think I should drop `usb` and just do `bluetooth`,`zeroconf`, and `dhcp` since those are the only ones that actually matter anyways

Reply:
I expect you are spot on about a custom component misusing the cache.  If we make it a util they will either fill it with garbage or if we make it a factory they could end up hold strong references to the object in the cache and leak memory if many or created.  Its just too easy to misuse so I removed the util 

So I made it private and limited it to the three integrations that it actually makes a difference in.

Reply:
We can do something like this to allow creating private instances:

```python
def create_memorized_fnmatch():
    @lru_cache(...)
    def memorized_fnmatch(...):
        ...

    return memorized_fnmatch
```